crates/fuel-core/src/graphql_api/storage/coins.rs

```diff
@@ -23,6 +41,180 @@ pub fn owner_coin_id_key(owner: &Address, coin_id: &UtxoId) -> OwnedCoinKey {
     default
 }
 
+/// The storage table for the index of coins to spend.
+
+// In the implementation of getters we use the explicit panic with the message (`expect`)
+// when the key is malformed (incorrect length). This is a bit of a code smell, but it's
+// consistent with how the `double_key!` macro works. We should consider refactoring this
+// in the future.
+pub struct CoinsToSpendIndex;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct CoinsToSpendIndexKey([u8; CoinsToSpendIndexKey::LEN]);
+
+impl Default for CoinsToSpendIndexKey {
+    fn default() -> Self {
+        Self([0u8; CoinsToSpendIndexKey::LEN])
+    }
+}
+
+impl core::fmt::Display for CoinsToSpendIndexKey {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(
+            f,
+            "owner={}, asset_id={}, retryable_flag={}, amount={}",
+            self.owner(),
+            self.asset_id(),
+            self.retryable_flag(),
+            self.amount()
+        )
+    }
+}
+
+impl CoinsToSpendIndexKey {
+    const LEN: usize = Address::LEN
+        + AssetId::LEN
+        + u8::BITS as usize / 8
+        + u64::BITS as usize / 8
+        + TxId::LEN
+        + 2;
+
+    #[allow(clippy::arithmetic_side_effects)]
+    pub fn from_coin(coin: &Coin) -> Self {
+        let address_bytes = coin.owner.as_ref();
+        let asset_id_bytes = coin.asset_id.as_ref();
+        let amount_bytes = coin.amount.to_be_bytes();
+        let utxo_id_bytes = utxo_id_to_bytes(&coin.utxo_id);
+
+        let mut arr = [0; CoinsToSpendIndexKey::LEN];
+        let mut offset = 0;
+        arr[offset..offset + Address::LEN].copy_from_slice(address_bytes);
+        offset += Address::LEN;
+        arr[offset..offset + AssetId::LEN].copy_from_slice(asset_id_bytes);
+        offset += AssetId::LEN;
+        arr[offset..offset + u8::BITS as usize / 8].copy_from_slice(&NON_RETRYABLE_BYTE);
```

Comment:
I think it will make more sense if the type of the coin `Message/Coin` will be after `amount`, before `Nonce/UtxoId`. In this case we also will sort messages by amount.

Reply:
This assumes we change the structure of the index key, because with the current approach we cannot switch places since this byte is used as a part of prefix when querying the index. Let's huddle this out.